File: Algorithms/25. Reverse Nodes in k-Group/kotlin/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseKGroup(head: ListNode?, k: Int): ListNode? {
        if (head == null) {
            return null
        }
        var reversed: ListNode? = null
        var current: ListNode? = head
        var tmp: ListNode?
        var count = 0

        while (count < k && current != null) {
            tmp = current?.next
            current.next = reversed
            reversed = current
            current = tmp
            count++
        }

        if (current == null && count < k) {
            return reverse(reversed)
        }

        head.next = reverseKGroup(current, k)
        return reversed
    }

    fun reverse(head: ListNode?): ListNode? {
        var reversed: ListNode? = null
        var current = head
        var tmp: ListNode?
        while (current != null) {
            tmp = current?.next
            current.next = reversed
            reversed = current
            current = tmp
        }
        return reversed
    }
}

File: Algorithms/25. Reverse Nodes in k-Group/swift/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func reverseKGroup(_ head: ListNode?, _ k: Int) -> ListNode? {
        if (head == nil) {
            return nil
        }
        if (k < 2) {
            return head
        }
        var dummy = ListNode(0)
        guard let head = head else {
            return nil
        }
        dummy.next = head
        var current = dummy

        while(current.next != nil) {
            var start = current.next
            var end = current.next
            var count = 1
            while (count < k && end != nil) {
                end = end!.next
                count += 1
            }

            if (end == nil) {
                break
            }
            var next = end!.next
            end!.next = nil
            
            var tCurr = start
            var prev:ListNode? = nil
            while (tCurr != nil) {
                let tNext = tCurr!.next
                tCurr!.next = prev
                prev = tCurr!
                tCurr = tNext
            }

            current.next = end
            start!.next = next
            current = start!
        }

        return dummy.next
    }
}